sdk/tools/pefixup.c

```diff
@@ -121,26 +127,99 @@ static int add_loadconfig(unsigned char *buffer, PIMAGE_NT_HEADERS nt_header)
     return 1;
 }
 
+static int driver_fixup(int mode, unsigned char *buffer, PIMAGE_NT_HEADERS nt_header)
+{
+    /* GNU LD just doesn't know what a driver is, and has notably no idea of paged vs non-paged sections */
+    for (unsigned i = 0; i < nt_header->FileHeader.NumberOfSections; i++)
+    {
+        PIMAGE_SECTION_HEADER Section = IMAGE_FIRST_SECTION(nt_header) + i;
+
+        /* Known sections which can be discarded */
+        if ((strncasecmp((char*)Section->Name, "INIT", 4) == 0)
+            /* Bugcheck code needs this, so don't discard it for kernels */
+            || ((strncasecmp((char*)Section->Name, ".rsrc", 4) == 0) && (mode != MODE_KERNEL)))
+        {
+            Section->Characteristics |= IMAGE_SCN_MEM_DISCARDABLE;
+            continue;
+        }
+
+        /* Known sections which can be paged */
+        if ((strncasecmp((char*)Section->Name, "PAGE", 4) == 0)
```

Comment:
`.edata`, `INIT`, `.rsrc` should also be here (even for the kernel)

Reply:
INIT and .rsrc are discardable, not sure about .edata.

Reply:
Windows' `ntoskrnl.exe`: `.edata` is pageable
![image](https://user-images.githubusercontent.com/961957/111520898-9705fd00-8769-11eb-8d3b-25a03358925b.png)

Reply:
OK, this is what is in now :

- Remove alignment bits that LD adds and that are not needed
- Remove write-access from .rsrc
- Remove initialized bit for code sections (LD adds it although it's not needed)
- Add the non-paged bit, except for PAGE, .rsrc and .edata

Reply:
Why link.exe does not put the alignment though?

Reply:
I'm not sure. My guess would be that it may be important for COFF, but PE doesn't care since it has the SectionAlignment value anyway.